### PR TITLE
Adding CloudBigtableIO logging for all write operations.

### PR DESCRIPTION
**Pull request by sduskis**
Adding the existing exception logging for CloudBigtableMultiTableWriteFn.
Adding Trace level logging for both CloudBigtableMultiTableWriteFn and CloudBigtableSingleTableWriteFn.

### CODE REVIEW

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 1]
```diff
@@ -583,6 +583,49 @@ public static Pipeline initializeForWrite(Pipeline p) {
...
+    protected final Logger LOG = LoggerFactory.getLogger(getClass());
+    protected CloudBigtableConfiguration config;
+    protected transient Connection conn;
+
+    public AbstractCloudBigtableTableWriteFn(CloudBigtableConfiguration config) {
+      this.config = config;
+    }
+
+    /**
+     * Creates a {@link Connection} and a {@link BufferedMutator}.
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: remove BufferedMutator

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 2]
```diff
@@ -583,6 +583,49 @@ public static Pipeline initializeForWrite(Pipeline p) {
...
+    public void finishBundle(DoFn<In, Out>.Context c) throws Exception {
+      LOG.debug("Closing the Bigtable connection.");
+      conn.close();
+    }
+
+    /**
+     * Logs the {@link Context} and the exception's
+     * {@link RetriesExhaustedWithDetailsException#getExhaustiveDescription()}.
+     */
+    protected void logExceptions(Context context, RetriesExhaustedWithDetailsException exception) {
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: The following two methods don't belong in this class, no? in a util class perhaps?

Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: or have them in RetriesExhaustedWithDetailsException

Comment (sduskis) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: We need a class level Log object to log the exception.  RetriesExhaustedWithDetailsException is an hbase class, and not something we can control.  I'd rather not have a Util class for 2 lines of code :)

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 3]
```diff
@@ -634,19 +677,11 @@ public void onException(RetriesExhaustedWithDetailsException exception,
...
-      LOG.warn("For context {}: exception occured during bulk writing: {}", context,
-        exception.getExhaustiveDescription());
-    }
-
-    /**
      * Performs an asyc mutation via {@link BufferedMutator#mutate(Mutation)}.
      */
     @Override
     public void processElement(ProcessContext context) throws Exception {
+      LOG.trace("Persisting {}", Bytes.toString(context.element().getRow()));
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: consider:
if (LOG.isTraceEnabled()) {
  ...
}

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java (hunk 2 above)
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: I fine with this.

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java (hunk 3 above)
Comment (sduskis) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: good point.

File: bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java [hunk 4]
```diff
@@ -715,18 +737,15 @@ public void processElement(ProcessContext context) throws Exception {
       KV<String, Iterable<Mutation>> element = context.element();
       try (Table t = conn.getTable(TableName.valueOf(element.getKey()))) {
         List<Mutation> mutations = Lists.newArrayList(element.getValue());
+        LOG.trace("Persisting {} elements to table {}", mutations.size(), element.getKey());
         Object results[] = new Object[mutations.size()];
         t.batch(mutations, results);
+        LOG.trace("Finished persisting {} elements to table {}", mutations.size(), element.getKey());
+      } catch (RetriesExhaustedWithDetailsException exception) {
+        logExceptions(context, exception);
```
Comment (kevinsi4508) [inline bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java]: do logExceptions() and retrowException() always go together?

**Comment by kevinsi4508** [comment]
LGTM

**Pull request merged**